Update Gradio to 3.23.0

space-nuko opened the pull request:
**Describe what this pull request is trying to achieve.**

Huge webui update, bumps Gradio to version 3.23.0

Impressions so far:
- Seems somewhat more responsive than 3.16
- Lots of UX issues (blank space, huge page margins causing squished elements and rampant text overflow, dropdown bugginess)
- Also a lot of UX improvements stemming from new event handlers (use `.release()` instead of `.change()`)

**Additional notes and description of your changes**

Changes made:
- Converted webui's `style.css` to SCSS for easier maintenance, also supported in extensions now (`style.scss`)
- Updated CSS to be closer to current Gradio version's UX
  + Increased dropdown list size, cut down on dropdown item padding
  + Adjustments to tool button sizes/borders
  + Fixes for quicksettings bar
  + Merged in auto's CSS changes from his 3.22 branch
- Added a CSS Hotload button to the UI, recompiles all webui/extension (S)CSS on the backend and injects it dynamically (lost on page reload)
- Changed quicksettings sliders to only apply values when mouse is released, greatly improves UX
- The `Custom Code` script now uses Gradio's new `gr.Code` element
- New config option to support Gradio's [theme API](https://gradio.app/theming-guide)
- Removed one of auto's hacks for preserving the selected gallery image (it addressed https://github.com/gradio-app/gradio/issues/2984)
- Removed another one of auto's hacks related to canvas resizing (it addressed https://github.com/gradio-app/gradio/issues/1721)

Known extension issues:
- dynamic-prompts extension in its current state causes infinite recursion, patch filed at https://github.com/adieyal/sd-dynamic-prompts/pull/316
- Tag Autocomplete extension needed changes, patch filed at https://github.com/DominikDoom/a1111-sd-webui-tagcomplete/pull/141
- Depth Library extension's canvas gets cut off on the right side due to the new page margins
- [Image Browser](https://github.com/AlUlkesh/stable-diffusion-webui-images-browser) extension causes massive slowdown, and quicksettings never load

Known bugs:
- ~~`img2img` is broken, I think due to the `submit_img2img` JavaScript not being compatible~~
- ~~Gallery lightbox modal doesn't work~~
- ~~`Send to img2img` is broken~~
- ~~Sending parameters from `PNG Info` is broken~~
- ~~Pasting `params.txt` with the arrow button is broken~~
- ~~`Show all pages` in Settings tab no longer works~~
- Weird styling in Extensions tab table

**Environment this was tested in**

 - OS: Windows
 - Browser: Chrome, Firefox Nightly
 - Graphics card: NVIDIA RTX 3090

**Screenshots or videos of your changes**

Old webui theme ported by AUTOMATIC1111 (current):
<img width="1901" alt="2023-03-21 16_59_37-Stable Diffusion - Chromium" src="https://user-images.githubusercontent.com/24979496/226739569-18c920cc-2c59-410c-9191-5fc91f9bf04a.png">

Monochrome (using Gradio's theme API):
<img width="1897" alt="2023-03-21 20_35_30-Stable Diffusion - Chromium" src="https://user-images.githubusercontent.com/24979496/226772286-47bc3908-3444-4e6b-a38b-f8c8bd00335f.png">

Gradio's default theme:
<img width="1905" alt="2023-03-20 00_39_18-Stable Diffusion - Chromium" src="https://user-images.githubusercontent.com/24979496/226249478-6b86fb81-c69c-4a33-a8cd-ab144c2f519f.png">

space-nuko committed:
Update Gradio to 3.22.1

space-nuko committed:
Support SASS compilation for system/user styles

space-nuko committed:
Update Gradio to ver. 3.22.1

space-nuko committed:
Fix notifications

space-nuko committed:
gradioApp().getRootNode() now returns document, append to body

space-nuko committed:
Fix gradio inputs hack

Seems another element was added to the arguments, so that pushed the
gallery image array back to the 4th from last element

space-nuko committed:
Fix img2img

Sakura-Luna commented:
I tried installing sass and found that it requires a native build install, and since it was last updated in 2014, apparently it's no longer maintained, I don't think introducing an outdated dependency that requires a build is a good option.

space-nuko commented:
2014? The one I installed from pip was last updated two weeks ago

https://github.com/sass/libsass-python

Maybe the lack of significant updates means it's just feature complete, regardless I think SCSS made it far easier to patch the stylesheets for this merge

Sakura-Luna commented:
On pypi, sass points to this https://pypi.org/project/sass/#history

space-nuko commented:
The one I added was named `libsass`

https://pypi.org/project/libsass/#history

Sakura-Luna commented:
It seems that it needs to join requirements.txt to avoid misunderstanding.

space-nuko commented:
I did put it there though

https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/8752/files#diff-2284b86f286dc7e0ea4bd09a0ec20c78fbb17d6724d7f0053e78428d0715bbb1R31

space-nuko committed:
Fix lightbox

space-nuko committed:
Quicksettings sliders now take effect on release only

space-nuko committed:
Fix blocks elements cutting off children that overflow

Tag autocomplete window, "Seed" label

space-nuko committed:
Fix extra networks metadata display

space-nuko committed:
Refresh highres calculator only when sliders are released

space-nuko committed:
Adjust size of img2img copy buttons

space-nuko committed:
Better UX for custom code script

space-nuko committed:
Fix prompt parsing from paste button

Gradio 3.22.1 seems to think a JS list of one element is itself the
single argument when passing it to the event handler.

Before: `["prompt"]` became `paste_func("prompt")`

Now: `["prompt"]` becomes `paste_func("['prompt']")`

So just return `"prompt"` instead

space-nuko committed:
Fix Send To buttons

Sakura-Luna commented:
Somewhat confused, I don't seem to notice a performance improvement, when I open the page after waiting for create ui to complete, it still looks like it's loading. By the way, this loading animation hopes to be blocked to keep it consistent with the official version.
![image](https://user-images.githubusercontent.com/53183413/226523974-6a86f0b5-5284-49c6-96c3-f048029402cb.png)

Looks like it still needs a lot of css fixes like rounded corners for some widgets, hyperlink colors and checkboxes etc.
![image](https://user-images.githubusercontent.com/53183413/226525126-b1af4972-16e7-41eb-9da2-1c8d2622ebc7.png)
![image](https://user-images.githubusercontent.com/53183413/226525303-6efd94c1-b491-49f0-b971-e7450faa8225.png)
![image](https://user-images.githubusercontent.com/53183413/226525007-563a05b6-61ee-4f54-a810-e7fe4cdd2ee8.png)

space-nuko committed:
Fix img2img aspect ratio overlay

space-nuko commented:
Yeah by "smoother" I meant less of a delay when for example selecting a script that injects a whole lot of UI elements onto the page, the quicksettings thing is weird and I don't know what's up with it

space-nuko committed:
Fix styling of tool button in quicksettings

space-nuko committed:
Fix aspect ratio overlay for img2img image case

hithereai commented:
Looks awful on my monitor (4k 42inch): 

![image](https://user-images.githubusercontent.com/121192995/226625506-2962e0ef-d7a3-4d61-804d-fc242c0755b5.png)
![image](https://user-images.githubusercontent.com/121192995/226625548-dc3d5e8f-6384-4e34-a4b5-d0c8200df045.png)
![image](https://user-images.githubusercontent.com/121192995/226625639-40c19be0-409e-408b-897c-9187e5949c19.png)

And I can keep going, basically all tabs are broken, visually.

Sakura-Luna commented:
> [gradio-app/gradio#3097](https://github.com/gradio-app/gradio/discussions/3097) better wait for the next release as gradio may introduce DOM changes to the components which possibly will break whatever changes made

@anapnoe I think you misunderstood the meaning, he refers to not controlling the consistency of css between versions, not "the next version".

anapnoe commented:
> > [gradio-app/gradio#3097](https://github.com/gradio-app/gradio/discussions/3097) better wait for the next release as gradio may introduce DOM changes to the components which possibly will break whatever changes made
> 
> @anapnoe I think you misunderstood the meaning, he refers to not controlling the consistency of css between versions, not "the next version".

The **DOM might change** when we release the next version, we do not version it nor can we guarantee its consistency  between versions 
as I understand it when DOM changes chances are css selectors become invalid for example consider tabs that have a kind of inefficient DOM structure tab buttons and tabitems are on the same level of selectors, in order to select the tab buttons 
`#tabs > div:first-child` is the only selector that works the id containing all the tab button is not existent css selector are  
`flex border-b-2 flex-wrap dark:border-gray-700` which is generic applied to other elements as well my point it if I want to improve the structure of tab component the first think I would do is to separate the tab-buttons and tab-items to do that I will have to create parent element for tab-items all the selectors that you refer them as div + div or div > div + div are prone to break when DOM changes and is hard to debug cause they work but for other target divs

anapnoe commented:
> > > [gradio-app/gradio#3097](https://github.com/gradio-app/gradio/discussions/3097) better wait for the next release as gradio may introduce DOM changes to the components which possibly will break whatever changes made
> > 
> > 
> > @anapnoe I think you misunderstood the meaning, he refers to not controlling the consistency of css between versions, not "the next version".
> 
> The **DOM might change** when we release the next version, we do not version it nor can we guarantee its consistency between versions as I understand it when DOM changes chances are css selectors become invalid for example consider tabs that have a kind of inefficient DOM structure tab buttons and tabitems are on the same level of selectors, in order to select the tab buttons `#tabs > div:first-child button` is the only selector that works the id containing all the tab buttons is not existent css selectors are `flex border-b-2 flex-wrap dark:border-gray-700` which is generic applied to other elements as well my point is if I wanted to improve the structure of tab component the first think I would do is to separate the tab-buttons and tab-items to do that I will have to create parent element for tab-items all the selectors that you refer them as div + div or div > div + div are prone to break when DOM changes and is hard to debug cause they work but for other target divs
I think gradio is doing a great work towards the right direction but it will take some time as all good things are

space-nuko commented:
> Looks awful on my monitor

Personally I don't mind the style changes as much (save for the numerous bugs), I also have a 4k monitor, 32-inch. I think this is what gradio's theming API is supposed to address isn't it? Maybe it'd be better to have a legacy and "default Gradio" theme integrated so people's workflows aren't impacted

> The DOM might change when we release the next version, we do not version it nor can we guarantee its consistency between versions

This has basically already happened as I was making the changes for 3.22.1, it wasn't a straightforward port. Again seems like the theme API is supposed to fix this by allowing custom styles to be attached

> https://github.com/gradio-app/gradio/discussions/3097#discussioncomment-5369369

What. I didn't know auto was working on 3.22 at the same time. Maybe we should collab?

space-nuko commented:
Updated to use auto's CSS, check the preview

space-nuko committed:
Merge in AUTOMATIC1111's 3.22.1 changes

space-nuko committed:
Don't reset padding for accordions

space-nuko committed:
Remove extra print

space-nuko committed:
Update Gradio to 3.23.0

Fixes a few dropdown issues

space-nuko committed:
Support Gradio theme API

space-nuko committed:
Make progress bar background obey current theme colors

space-nuko committed:
Adjust selected dropdown item color for better visiblity with themes

Sakura-Luna commented:
After testing, the [image browser](https://github.com/AlUlkesh/stable-diffusion-webui-images-browser.git) extension will keep the page stuck loading.

space-nuko committed:
Fix aspect ratio overlay event when page is still loading

space-nuko committed:
Fix infinite recursion in gradioApp()

space-nuko commented:
Fixed an issue that caused infinite recursion with that extension, but the performance on page load is still very bad. I'm noticing it's trying to load a bunch of images from my outputs folder on startup for some reason.

space-nuko commented:
@Sakura-Luna how many rows/columns do you have configured for the image browser in your settings? It started working again for me when I put 6/6. (The "Minimum number of pages per load" option doesn't seem to do anything in that fork)
Probably a performance issue with new Gradio at this point.

Sakura-Luna commented:
In my case it's still loading all the time, 6/6 gradio 3.22.1.

Cyberbeing commented:
The temporary file behavior seems even worse in Gradio 3.23.0 than it was in 3.16.2.

Similar to 3.16.2, Gradio 3.23.0 still ignores `Directory for temporary images` in settings, but now it also creates a bunch of empty folders in %TEMP% on startup, while also placing every saved image into a separate folder next to a 0 byte tmp file. If you are able to fix both these issues in your pull request and force Gradio to use a single temp folder (default or specified), that would be great.

![gradio_emptytmp](https://user-images.githubusercontent.com/917535/227187973-eed05d5f-5a22-4f00-8caa-00ee7834b482.png)

![gradiosavedtmp](https://user-images.githubusercontent.com/917535/227188773-c5d3d2b6-5f1c-4ed7-88e1-a8e660f67b81.png)

Cyberbeing commented:
Another regression from 3.16.2 I've noticed, is when clicking an image to view it zoomed in during generation, the zoomed in image does not refresh with the new image when generation finishes.

Sakura-Luna commented:
> Another regression from 3.16.2 I've noticed, is when clicking an image to view it zoomed in during generation, the zoomed in image does not refresh with the new image when generation finishes.

This is not a problem introduced by gradio 3.23. The current version exists https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/8162, and it has nothing to do with zooming in. I think it may be a program logic error.

Cyberbeing commented:
> This is not a problem introduced by gradio 3.23. The current version exists #8162, and it has nothing to do with zooming in. I think it may be a program logic error.

The issue you linked is something different than the problem I just described with gradio 3.23. What I'm describing is:
* Click Generate
* Click current image result in txt2img_gallery to view it zoomed in while Generate is in progress to view it zoomed in
* Generate finishes and refreshes the image in Result window, but not the current zoomed view

With 3.16.2:
* Click Generate
* Click current image result in txt2img_gallery to view it zoomed in while Generate is in progress to view it zoomed in
* Generate finishes and refreshes _both_ the image in Result window and the current zoomed view

Sometimes that automatic zoomed refresh behavior does break in 3.16.2, but usually it works. When testing 3.23 the zoomed view _never_ refreshes until you close and reopen the zoomed view which is a regression.

The issue you linked sounds like a unrelated problem.I've never had a problem with the result window not refreshing in either 3.16.2 or 3.23.

space-nuko commented:
> The temporary file behavior seems even worse in Gradio 3.23.0 than it was in 3.16.2.

Seems like this PR might address some of those problems: https://github.com/gradio-app/gradio/pull/3523

space-nuko committed:
Fix lightbox not updating on new generation finish

Cyberbeing commented:
It's not image previews which end up in the temp folder, but rather files served as downloads whenever you hit the SAVE or ZIP buttons. So ultimately what `Directory for temporary images` should be doing is overriding the directory which Gradio serves image/zip file downloads from.

Currently Graido always dumps temporary files (images/zips you save) into `%TEMP%` on Windows which is never cleaned up.
* Generate Images
* Press Save/Zip
* Gradio Dump your Image/Zip into `%TEMP%` which it then serves as a download
* These temp files are never cleaned up automatically, resulting in rather extremely storage usage and I/O overhead if you save a lot of files

As far as I can tell, this problem of Gradio ignoring custom temp folders may have started webui migrated to Gradio 3.16.2. The webui code for `Directory for temporary images` has no effect on current Gradio.

For the webui use-case this behavior is very inefficient, since for single images in particular there is no reason to duplicate them in `%TEMP%` when they already exist in `/output/` (i.e. the same image displayed as in txt2img_gallery could be directly served as a download, but Gradio seems unaware these files exist). In fact, if you just right click save as on the gallery result image, rather than click the save button you avoid this temp file problem entirely. Unfortunately, no way to avoid temp files with saved zips since they are generated on the fly.

If the `Directory for temporary images` setting were working, you'd be able to specify a custom temp folder for Gradio to prevent it from dumping any temp files or folders into system `%TEMP%`. This would then allow you to use the webui option to automatically clean your custom temp folder on startup.

space-nuko commented:
Thanks for clarifying, filed a Gradio feature request https://github.com/gradio-app/gradio/issues/3597

space-nuko commented:
@Sakura-Luna 

> In my case it's still loading all the time, 6/6 gradio 3.22.1.

I think it's because that fork has some kind of weird concurrency issue in the JavaScript with latest Gradio, try https://github.com/nightglimmer/stable-diffusion-webui-images-browser/network instead?

AlUlkesh commented:
@Sakura-Luna 

> In my case it's still loading all the time, 6/6 gradio 3.22.1.

I just made a new install with this pr. I don't see any slowdown on the image browser. This is on Gradio 3.23.0.

Was this perhaps the first time you used a newer version of the browser on a very large directory? On the very first access of an image directory it will populate the database with the exif info of every single image, which can take a while. But after that it should be fast.

space-nuko commented:
That freezing/hanging reproduces for me only if I turn up the number of rows/columns in the settings sufficiently high. Otherwise it works normally

Could it have to do with https://github.com/AlUlkesh/stable-diffusion-webui-images-browser/issues/91? Might be worth comparing the differences between both forks. I noticed there was some kind of async locking mechanism added recently. Could be a race condition/deadlock? https://github.com/nightglimmer/stable-diffusion-webui-images-browser/compare/main...AlUlkesh:stable-diffusion-webui-images-browser:main

Sakura-Luna commented:
@AlUlkesh This is weird, my new instance only has three images. I cleaned up the output path and configured json, disabled other extensions and reinstalled the image browser, but the problem still persists. Maybe I could try a different location?

AlUlkesh commented:
@Sakura-Luna , @space-nuko 

I tried with 15x15, Still no slowdown. However, the initial image  covers half the screen now. Also quicksettings / changing settings work.

UPDATE: I'm normally running on Firefox. I just tried Chrome and sure enough I can see the problem there. Which browsers did you test?

Hilurex commented on modules/ui.py:
```diff
@@ -15,12 +15,14 @@
 import numpy as np
 from PIL import Image, PngImagePlugin
 from modules.call_queue import wrap_gradio_gpu_call, wrap_queued_call, wrap_gradio_call
+import sass
```
![image](https://user-images.githubusercontent.com/104232913/227614937-a47bc322-b20c-4d6c-a332-2f6eee6451ab.png)

Hilurex replied:
not problem is am not install libsass on my pytom library XD, ty for help

Hilurex replied:
I already managed to execute it but it does not let me generate images since the generate button does not work

space-nuko commented:
I used Chrome (ungoogled-chromium) for my test, will try Firefox when I can

Sakura-Luna commented:
@AlUlkesh Stuck on Edge(chromium).

AUTOMATIC1111 commented:
There are a lot of changes and they are mixed not only with each other as in other big PRs but this time also with my code from my branch. The preferred way to add code is to do small PRs. As it is, it's a hell to review.

SCSS will most likely not make it in regardless unless there is a very good argument for it.

AlUlkesh commented:
@Sakura-Luna , @space-nuko
I changed the image browser's javascript. At least for me, nothing seems stuck anymore, also on Chrome.

There's still some visual issues with the gallery, on 15x15 the first thumbnail display is giant - but at least now the main webui seems to not be affected anymore.

anapnoe commented:
> There are a lot of changes and they are mixed not only with each other as in other big PRs but this time also with my code from my branch. The preferred way to add code is to do small PRs. As it is, it's a hell to review.
> 
> SCSS will most likely not make it in regardless unless there is a very good argument for it.
> 
> Otherwise, thanks for the extension compatibility research. The patch for Tag Autocomplete will make it work but the event for element change will likely not fire because now it's just a simple `span` rather than a `select`

also it fires twice it is not a select input field any more but a list ul/li it dispatches click events one for open and one for select - close, I don't like this change it is messy it calls the _js function twice so you need to be extra careful I haven't test, if it does the same for the py function but I guess it would do the same thing as they are bound together if someone can test it to confirm if it respects at least the py function this can cause a lot of bugs for extensions and maybe delays for intensive tasks like checkpoint loading etc.

wzgrx commented:
No matching distribution found for gradio==3.23 ？

wzgrx commented:
File "D:\Soft\Aidrw\stable-diffusion-webui\launch.py", line 326, in prepare_environment
    run_pip(f"install -r \"{requirements_file}\"", "requirements for Web UI")
  File "D:\Soft\Aidrw\stable-diffusion-webui\launch.py", line 145, in run_pip
    return run(f'"{python}" -m pip {args} --prefer-binary{index_url_line}', desc=f"Installing {desc}", errdesc=f"Couldn't install {desc}")
  File "D:\Soft\Aidrw\stable-diffusion-webui\launch.py", line 113, in run
    raise RuntimeError(message)
RuntimeError: Couldn't install requirements for Web UI.
提示：Python 运行时抛出了一个异常。请检查疑难解答页面。
Command: "D:\Soft\Aidrw\stable-diffusion-webui\Python3.10\python.exe" -m pip install -r "requirements_versions.txt" --prefer-binary --index-url https://mirrors.bfsu.edu.cn/pypi/web/simple
Error code: 1
stdout: Looking in indexes: https://mirrors.bfsu.edu.cn/pypi/web/simple, https://pypi.tuna.tsinghua.edu.cn/simple, https://mirrors.pku.edu.cn/pypi/simple, https://mirror.nju.edu.cn/pypi/web/simple, https://mirror.sjtu.edu.cn/pypi/web/simple, https://pypi.doubanio.com/simple

Looking in links: https://mirror.sjtu.edu.cn/pytorch-wheels/torch_stable.html

Requirement already satisfied: blendmodes==2022 in d:\soft\aidrw\stable-diffusion-webui\python3.10\lib\site-packages (from -r requirements_versions.txt (line 1)) (2022)

Collecting transformers==4.25.1

  Using cached https://pypi.doubanio.com/packages/90/5b/02becd3e137fdacdc30c6442d04094d0b92e500f978df00b1a197b68fd2b/transformers-4.25.1-py3-none-any.whl (5.8 MB)

Collecting accelerate==0.12.0

  Using cached https://pypi.doubanio.com/packages/52/36/1a3aec552da693acbb65f9f3613433bea0e065448ef27dc7a9c2f2fb1efa/accelerate-0.12.0-py3-none-any.whl (143 kB)

Requirement already satisfied: basicsr==1.4.2 in d:\soft\aidrw\stable-diffusion-webui\python3.10\lib\site-packages (from -r requirements_versions.txt (line 4)) (1.4.2)

Requirement already satisfied: gfpgan==1.3.8 in d:\soft\aidrw\stable-diffusion-webui\python3.10\lib\site-packages (from -r requirements_versions.txt (line 5)) (1.3.8)

stderr: ERROR: Could not find a version that satisfies the requirement gradio==3.23 (from versions: 0.1.0, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5, 0.1.6, 0.1.7, 0.1.8, 0.1.9, 0.2.0, 0.2.1, 0.3.0, 0.3.1, 0.3.2, 0.3.3, 0.3.4, 0.3.5, 0.4.0, 0.4.1, 0.4.2, 0.4.4, 0.5.0, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.7.4, 0.7.5, 0.7.6, 0.7.7, 0.7.8, 0.8.0, 0.8.1, 0.9.0, 0.9.1, 0.9.2, 0.9.3, 0.9.4, 0.9.5, 0.9.6, 0.9.7, 0.9.8, 0.9.9.2, 0.9.9.3, 0.9.9.5, 0.9.9.6, 0.9.9.7, 0.9.9.8, 0.9.9.9, 0.9.9.9.2, 1.0.0a1, 1.0.0a3, 1.0.0a4, 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.0.5, 1.0.6, 1.0.7, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.1.4, 1.1.5, 1.1.6, 1.1.8, 1.1.8.1, 1.1.9, 1.2.2, 1.2.3, 1.3.0, 1.3.1, 1.3.2, 1.4.0, 1.4.2, 1.4.3, 1.4.4, 1.5.0, 1.5.1, 1.5.3, 1.5.4, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.7.0, 1.7.1, 1.7.2, 1.7.3, 1.7.4, 1.7.5, 1.7.6, 1.7.7, 2.0.0, 2.0.1, 2.0.2, 2.0.4, 2.0.5, 2.0.6, 2.0.7, 2.0.8, 2.0.9, 2.0.10, 2.1.0, 2.1.1, 2.1.2, 2.1.4, 2.1.6, 2.1.7, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.2.6, 2.2.7, 2.2.8, 2.2.9a0, 2.2.9a2, 2.2.10, 2.2.11, 2.2.12, 2.2.13, 2.2.14, 2.2.15, 2.3.0a0, 2.3.0b99, 2.3.0b101, 2.3.0b102, 2.3.0, 2.3.3, 2.3.4, 2.3.5b0, 2.3.5, 2.3.6, 2.3.7b0, 2.3.7b1, 2.3.7b2, 2.3.7, 2.3.8b0, 2.3.9, 2.4.0a0, 2.4.0, 2.4.1, 2.4.2, 2.4.4, 2.4.5, 2.4.6, 2.4.7b0, 2.4.7b2, 2.4.7b3, 2.4.7b4, 2.4.7b5, 2.4.7b6, 2.4.7b7, 2.4.7b8, 2.4.7b9, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.5.8a0, 2.6.0, 2.6.1a0, 2.6.1b0, 2.6.1b3, 2.6.1, 2.6.2, 2.6.3, 2.6.4b0, 2.6.4b2, 2.6.4b3, 2.6.4, 2.7.0a101, 2.7.0a102, 2.7.0b70, 2.7.0, 2.7.5, 2.7.5.1, 2.7.5.2b0, 2.7.5.2, 2.8.0a100, 2.8.0b0, 2.8.0b2, 2.8.0b3, 2.8.0b4, 2.8.0b5, 2.8.0b6, 2.8.0b10, 2.8.0b12, 2.8.0b20, 2.8.0b22, 2.8.0, 2.8.1, 2.8.2, 2.8.3, 2.8.4, 2.8.5, 2.8.6, 2.8.7, 2.8.8, 2.8.9, 2.8.10, 2.8.11, 2.8.12, 2.8.13, 2.8.14, 2.9.0b0, 2.9.0b1, 2.9.0b2, 2.9.0b3, 2.9.0b5, 2.9.0b6, 2.9.0b7, 2.9.0b8, 2.9.0b9, 2.9.0b10, 2.9b11, 2.9b12, 2.9b13, 2.9b14, 2.9b15, 2.9b20, 2.9b21, 2.9b22, 2.9b23, 2.9b24, 2.9b25, 2.9b26, 2.9b27, 2.9b28, 2.9b30, 2.9b31, 2.9b32, 2.9b33, 2.9b40, 2.9b48, 2.9b50, 2.9.0, 2.9.0.1, 2.9.1, 2.9.2, 2.9.3, 2.9.4, 3.0b0, 3.0b1, 3.0b2, 3.0b5, 3.0b6, 3.0b8, 3.0b9, 3.0b10, 3.0, 3.0.1b120, 3.0.1b121, 3.0.1b300, 3.0.1, 3.0.2, 3.0.3, 3.0.4, 3.0.5, 3.0.6b1, 3.0.6b2, 3.0.6b3, 3.0.6, 3.0.7, 3.0.8b1, 3.0.8, 3.0.9b10, 3.0.9b11, 3.0.9b20, 3.0.9, 3.0.10b2, 3.0.10b16, 3.0.10, 3.0.11b1, 3.0.11, 3.0.12, 3.0.13b13, 3.0.13b15, 3.0.13b100, 3.0.13, 3.0.14, 3.0.15, 3.0.16, 3.0.17, 3.0.18b0, 3.0.18, 3.0.19b0, 3.0.19b1, 3.0.19b2, 3.0.19, 3.0.20.dev0, 3.0.20, 3.0.21, 3.0.22, 3.0.23.dev1, 3.0.23, 3.0.24, 3.0.25, 3.0.26, 3.1.0, 3.1.1, 3.1.2, 3.1.3a0, 3.1.3a2, 3.1.3a3, 3.1.3a4, 3.1.3a5, 3.1.3, 3.1.4b0, 3.1.4b1, 3.1.4b2, 3.1.4b3, 3.1.4b4, 3.1.4b5, 3.1.4, 3.1.5b1, 3.1.5b2, 3.1.5b3, 3.1.5b4, 3.1.5b5, 3.1.5b7, 3.1.5b8, 3.1.5b9, 3.1.5b10, 3.1.5, 3.1.6b1, 3.1.6, 3.1.7, 3.1.8b0, 3.1.8b2, 3.1.8b3, 3.1.8b4, 3.1.8b6, 3.2, 3.2.1b0, 3.2.1b1, 3.2.1b2, 3.3b0, 3.3b1, 3.3, 3.3.1, 3.4b0, 3.4b1, 3.4b2, 3.4b3, 3.4b5, 3.4, 3.4.1, 3.5, 3.6.0b1, 3.6.0b2, 3.6.0b3, 3.6.0b7, 3.6.0b10, 3.6, 3.7, 3.8b1, 3.8b2, 3.8, 3.8.1.dev1, 3.8.1, 3.8.2, 3.9, 3.9.1, 3.10.0, 3.10.1, 3.11.0, 3.12.0b1, 3.12.0b2, 3.12.0b3, 3.12.0b6, 3.12.0b7, 3.12.0, 3.13.0b1, 3.13.0, 3.13.1b0, 3.13.1b1, 3.13.1b2, 3.13.1, 3.13.2, 3.14.0a1, 3.14.0, 3.15.0, 3.16.0, 3.16.1b1, 3.16.1, 3.16.2, 3.17.0, 3.17.1b1, 3.17.1b2, 3.17.1, 3.18.0, 3.18.1b1, 3.18.1b2, 3.18.1b3, 3.18.1b4, 3.18.1b5, 3.18.1b6, 3.18.1b7, 3.19.0, 3.19.1, 3.20.0b1, 3.20.0b2, 3.20.0, 3.20.1, 3.21.0, 3.22.0, 3.22.1b1, 3.22.1)

ERROR: No matching distribution found for gradio==3.23

anapnoe commented:
> > There are a lot of changes and they are mixed not only with each other as in other big PRs but this time also with my code from my branch. The preferred way to add code is to do small PRs. As it is, it's a hell to review.
> > SCSS will most likely not make it in regardless unless there is a very good argument for it.
> > Otherwise, thanks for the extension compatibility research. The patch for Tag Autocomplete will make it work but the event for element change will likely not fire because now it's just a simple `span` rather than a `select`
> 
> also it fires twice it is not a select input field any more but a list ul/li it dispatches click events one for open and one for select - close, I don't like this change it is messy it calls the _js function twice so you need to be extra careful I haven't test, if it does the same for the py function but I guess it would do the same thing as they are bound together if someone can test it to confirm if it respects at least the py function this can cause a lot of bugs for extensions and maybe delays for intensive tasks like checkpoint loading etc.
> 
> also Auto you shouldn't have upgrated the main branch, you should have kept it separate for people to test and see if everything works ok, there are a ton of issues addressed by @space-nuko on gradio bugs post that will also become part of the ui did you checked them?



> No matching distribution found for gradio==3.23 ？

maybe it should be like this 3.23.0

anapnoe commented:
> No matching distribution found for gradio==3.23 ？

maybe it should be like this 3.23.0

wzgrx commented:
> 3.23.0

ERROR: No matching distribution found for gradio==3.23.0

not work too

Sakura-Luna commented:
@wzgrx Remove the tsinghua source (and others) from your pip, it's outdated.

wzgrx commented:
> @wzgrx Remove the tsinghua source (and others) from your pip, it's outdated.

how to delete？

Sakura-Luna commented:
@wzgrx Search for `pip remove index url` in a search engine, it speaks more clearly than I do.

hithereai commented:
Webui has been updated with Gradio 3.23, this PR can be closed (?)

wzgrx commented:
> @wzgrx Search for `pip remove index url` in a search engine, it speaks more clearly than I do.

ok， tks

hithereai commented:
See https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/b0b777e64da238f1b259b77b0899b61c26a99dee

This PR is outdated as far as I can tell

space-nuko commented:
Huh? Okay. Guess I'll rebase...

space-nuko commented:
Looks like auto missed several fixes to Gradio changes like https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/8921 that were in this PR
